custom-elements/custom-elements-registry/define.html

```diff
@@ -0,0 +1,178 @@
+  // then throw a NotSupportedError and abort these steps.
+  var testDupCtor = class {};
+  test(function () {
+    window.customElements.define("test-define-dup-ctor", testDupCtor);
+  }, "test-define-dup-ctor 1st, success");
+  test(function () {
+    assert_throws(new NotSupportedError, function () {
+      window.customElements.define("test-define-dup-ctor2", testDupCtor);
+    });
+  }, "test-define-dup-ctor 2nd, throw a NotSupportedError");
+
+  // 7.1. If extends is a valid custom element name,
+  // then throw a NotSupportedError.
+  validCustomElementNames.forEach(function (name) {
+    test(function () {
+      assert_throws(new NotSupportedError, function () {
+        window.customElements.define("test-define-extend-valid-name", class {}, { extends: name });
+      });
+    }, 'If extends is ' + name + ', throw a NotSupportedError');
+  });
+
+  // 7.2. If the element interface for extends and the HTML namespace is HTMLUnknownElement
+  // (e.g., if extends does not indicate an element definition in this specification),
+  // then throw a NotSupportedError.
+  [
+    // https://html.spec.whatwg.org/multipage/dom.html#elements-in-the-dom:htmlunknownelement
+    "bgsound",
+    "blink",
+    "isindex",
+    "multicol",
+    "nextid",
+    "spacer",
+    "elementnametobeunknownelement",
+  ].forEach(function (name) {
+    test(function () {
+      assert_throws(new NotSupportedError, function () {
+        window.customElements.define("test-define-extend-" + name, class {}, { extends: name });
+      });
+    }, 'If extends is ' + name + ', throw a NotSupportedError');
+  });
+
+  // 10. Let prototype be Get(constructor, "prototype"). Rethrow any exceptions.
+  // 11. If Type(prototype) is not Object,
+  // then throw a TypeError exception.
+  test(function () {
+    var c = class {};
+    c.prototype = "string";
+    assert_throws(new TypeError, function () {
+      window.customElements.define("test-define-ctor-prototype-string", c);
+    });
+  }, 'If Type(constructor.prototype) is not Object, throw a TypeError');
+
+  // 12. Let connectedCallback be Get(prototype, "connectedCallback"). Rethrow any exceptions.
+  // 13. If connectedCallback is not undefined, and IsCallable(connectedCallback) is false,
+  // then throw a TypeError exception.
+  // 14. Let disconnectedCallback be Get(prototype, "disconnectedCallback"). Rethrow any exceptions.
+  // 15. If disconnectedCallback is not undefined, and IsCallable(disconnectedCallback) is false,
+  // then throw a TypeError exception.
+  // 16. Let attributeChangedCallback be Get(prototype, "attributeChangedCallback"). Rethrow any exceptions.
+  // 17. If attributeChangedCallback is not undefined, and IsCallable(attributeChangedCallback) is false,
+  // then throw a TypeError exception.
+  [ "connectedCallback", "disconnectedCallback", "attributeChangedCallback" ].forEach(function (name) {
+    test(function () {
+      var c = class {};
+      c.prototype[name] = {};
```

Comment:
Another good test would be setting it to `null`, which should throw, and setting it to `undefined`, which should not throw. (The difference between setting it to undefined and just leaving it alone is that when you set it to undefined, `c.prototype.hasOwnProperty("connectedCallback")` will be true.)